Jetpack Error UX: Add Jetpack connection error messages

wojtekn opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes:
- https://github.com/Automattic/dotcom-forge/issues/3097
- https://github.com/Automattic/dotcom-forge/issues/3095
- https://github.com/Automattic/dotcom-forge/issues/3408
- https://github.com/Automattic/dotcom-forge/issues/3455
- https://github.com/Automattic/dotcom-forge/issues/3407

## Proposed Changes

In this PR, I propose to add an error message for Jetpack connection health getting a PHP fatal error from the site. I also add tests for all three component states.

![jetpack-dns](https://github.com/Automattic/wp-calypso/assets/727413/c3f670b1-4da5-4a22-a61f-44c6d9b8c875)

![jetpack-token](https://github.com/Automattic/wp-calypso/assets/727413/f8a25303-88c9-420d-8657-456de37c9927)

![jetpack-fatal](https://github.com/Automattic/wp-calypso/assets/727413/0560195b-f2f4-4fcf-850d-7625efe27505)

![jetpack-generic](https://github.com/Automattic/wp-calypso/assets/727413/473f220e-61e7-4686-ab17-ae20d9401c80)

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Fatal error on the site

1. Create an Atomic site with a Business plan and activate hosting features
2. SSH to the site and cd to your site root
3. Edit the wp-config.php file and add the code:
```
fatal(doit);
```
4. Navigate to the Posts page in Calypso.
5. Confirm that the error message appears
6. Confirm that the "Learn how to fix" link points to the proper page in the documentation

### Token error

1. Create an Atomic site with a Business plan and activate hosting features
2. SSH to the site and cd to your site root
3. Run:
```
wp option patch delete jetpack_private_options user_tokens
```
4. Navigate to the Posts page in Calypso.
5. Confirm that the error message appears
6. Confirm that the "Learn how to fix" link points to the proper page in the documentation

### Site does not resolve HTTP

1. Create an external WP site
2. Connect jetpack
3. SSH to the site and cd to your site root
4. Edit `index.php` and add code that triggers timeout:
```
sleep(60);
```
5. Navigate to the Posts page in Calypso.
6. Confirm that the error message appears
7. Confirm that the "Learn how to fix" link points to the proper page in the documentation

### Generic message for non-admin user

1. Create an Atomic site with a Business plan and activate hosting features
2. Add non-admin user
3. SSH to the site and cd to your site root
4. Edit the wp-config.php file and add the code:
```
fatal(doit);
```
5. Logged as a non-admin user, navigate to the Posts page in Calypso.
6. Confirm that the generic error message appears
7. Confirm that there is no "Learn how to fix" link

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

wojtekn committed:
Add Jetpack connection fatal error message

wojtekn committed:
Add tests for three existing component states

danielbachhuber committed:
Merge branch 'trunk' into add/jetpack-connection-fatal-error-message

danielbachhuber reviewed (commented):
Works as expected 👍 

I don't love the error message, though. I think we should draft all of them in a Google Doc, wordsmith them, and then add them to Calypso.

wojtekn committed:
Add all constants used by API endpoint

wojtekn committed:
Add support for Jetpack token error message

wojtekn committed:
Add HTTP connection error message

wojtekn committed:
Ensure that errorType is always correct

wojtekn committed:
Add support for Jetpack connection error caused by missing plugin

wojtekn committed:
Add support for generic error for non-admin user

wojtekn committed:
Define props as optional

wojtekn committed:
Reduce repetition and fix type handling

wojtekn commented:
@danielbachhuber thanks for reviewing it.

> I don't love the error message, though. I think we should draft all of them in a Google Doc, wordsmith them, and then add them to Calypso.

I've added support for all cases and added all messages to Google Docs.

wojtekn committed:
Check for CTA in tests

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -45,6 +56,75 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+			<ErrorNotice
+				errorType={ errorType }
+				errorText={ translate(
+					"Jetpack is unable to communicate with your site due to a critical error that has occurred there. Please check your site admin's email inbox for instructions."
```
```suggestion
					"Jetpack is unable to communicate with your site due to a critical error on the site."
```

I think we should drop the "Please check your site admin's email inbox" for a few reasons:
1. We're already directing them to 'Learn how to fix'
2. They may or may not have access to the site admin's inbox.
3. Depending on the fatal, the email may or may not actually be sent. For example, with the fatal in `wp-config.php`, the email isn't sent because the fatal error trapper isn't registered yet.

Also worth noting: for WordPress.com on Atomic sites, we can point them to the 'PHP logs' tab on the new 'Site Monitoring' page for visibility into what error is.

javierarce replied:
Instead of saying `Jetpack is unable to communicate with your site due to a critical error on the site.` I'd simply say: `Jetpack can't communicate with your site due to a critical error.`

wojtekn replied:
As we agreed on Slack, let's leave the current one to say where the fatal error occurred clearly.

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -45,6 +56,75 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+			<ErrorNotice
+				errorType={ errorType }
+				errorText={ translate(
+					'Jetpack is unable to communicate with your site due to a token error. Please reconnect Jetpack.'
```
```suggestion
					'Jetpack is unable to communicate with your site because your site isn’t connected.'
```

I think 'token error' falls into the category of '[technical jargon](https://twitter.com/carlvellotti/status/1688714484175589376/photo/1)'. Also, 'Please reconnect Jetpack' conflicts with the notice action text.

javierarce replied:
+1 but instead of `Jetpack is unable to communicate` I'd say `Jetpack can't communicate`

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -45,6 +56,75 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+				noticeActionHref={ localizeUrl(
+					'https://wordpress.com/support/why-is-my-site-down/#theres-an-issue-with-your-sites-jetpack-connection'
+				) }
+				noticeActionText={ translate( 'Learn how to fix' ) }
```
```suggestion
				noticeActionText={ translate( 'Learn how to reconnect Jetpack' ) }
```

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -45,6 +56,75 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+			<ErrorNotice
+				errorType={ errorType }
+				errorText={ translate(
+					'Jetpack is unable to communicate with your site due to a HTTP connection error. Please ensure that your site serves requests.'
```
```suggestion
					'Jetpack is unable to communicate with your site because your site isn’t responding to requests.'
```

I think 'HTTP connection error' is too jargony. Also, 'Please ensure that your site serves requests' is better left for the support doc.

javierarce replied:
Like in my previous comments, let's say `Jetpack can't communicate…`

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -45,6 +56,75 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+			<ErrorNotice
+				errorType={ errorType }
+				errorText={ translate(
+					'Jetpack is unable to communicate with your site due to an inactive Jetpack plugin. Please ensure that Jetpack plugin is active and reconnect Jetpack.'
```
```suggestion
					'We are unable to communicate with your site because the Jetpack plugin is deactivated.'
```

It's a little confusing to say that 'Jetpack is unable to communicate with your site because it's deactivated...'

wojtekn replied:
Jetpack can mean different things, should we say `[...] because Jetpack plugin is deactivated`?

danielbachhuber replied:
@wojtekn Sure, I updated my suggestion.

javierarce replied:
Let's say `We can't communicate…` instead of `We are unable to communicate…`

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -45,6 +56,75 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+				noticeActionHref={ localizeUrl(
+					'https://wordpress.com/support/why-is-my-site-down/#theres-an-issue-with-your-sites-jetpack-connection'
+				) }
+				noticeActionText={ translate( 'Learn how to fix' ) }
```
```suggestion
				noticeActionText={ translate( 'Learn how to reactivate Jetpack' ) }
```

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -45,6 +56,75 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+			<ErrorNotice
+				errorType={ errorType }
+				errorText={ translate(
+					'Jetpack is unable to communicate with your site. Please contact site administrator.'
```
```suggestion
					'Jetpack is unable to communicate with your site.'
```

danielbachhuber replied:
Do we need this `if ( errorType === GENERIC_ERROR ) {` branch if it's essentially duplicative of `UNKNOWN_ERROR` below?

wojtekn replied:
`UNKNOWN_ERROR` would appear when a new error class is introduced server-side or when the connection is not working properly, but we couldn't match the problem to any known case. `GENERIC_ERROR` would be shown when the user is not an admin.

Do you think we don't need to differentiate messaging, e.g., ask users to contact support in one case and ask them to contact site admin in the other?

javierarce replied:
Like in my previous comments, let's say `Jetpack can't communicate…`

danielbachhuber reviewed (commented):
Great start!

I think it'd be good to get @javierarce and potentially @Automattic/reactor's perspectives on these messages too.

wojtekn committed:
Reword error messages and CTAs

wojtekn committed:
Clarify error message means Jetpack plugin

wojtekn committed:
Update tests to match implementation

javierarce commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -30,10 +39,12 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
-				errorType={ DNS_ERROR }
+				errorType={ errorType }
 				errorText={ translate(
 					"Jetpack is unable to connect to your domain. Please update your domain's DNS records so they're pointed properly to your site."
```
Let's say: `Jetpack can't connect to your domain. Ensure your domain's DNS records are correctly pointing to your site.`

wojtekn committed:
Improve error message wording

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -30,12 +39,14 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+				errorType={ errorType }
 				errorText={ translate(
-					"Jetpack is unable to connect to your domain. Please update your domain's DNS records so they're pointed properly to your site."
+					'Jetpack is unable to connect to your domain. Ensure your domain’s DNS records are correctly pointing to your site.'
```
```suggestion
					'Jetpack is unable to connect to your domain because your domain’s DNS records aren’t pointing to your site.'
```

This suggestion keeps the message focused on the error, instead of starting to introduce the solution.

danielbachhuber reviewed (approved):
Just one other suggested text tweak. The others look good to me.

wojtekn committed:
Update wording to avoid introducing a solution

wojtekn committed:
Update test to follow UX change

wojtekn committed:
Merge branch 'trunk' into add/jetpack-connection-fatal-error-message

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8714246

Hi @wojtekn, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Jetpack can’t communicate with your site due to a critical error on the site.`
- `Jetpack can’t communicate with your site because your site isn’t connected.`
- `Learn how to reconnect Jetpack`
- `Jetpack can’t communicate with your site because your site isn’t responding to requests.`
- `We can’t communicate with your site because the Jetpack plugin is deactivated.`
- `Learn how to reactivate Jetpack`
- `Jetpack can’t communicate with your site. Please contact site administrator.`
- `Jetpack can’t communicate with your site.`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8714246).
